### DISCUSSION TITLE
Need example or solution idea about GetMiddleware implementation with FirebaseAuth

### DISCUSSION THREAD

**Discussion by furatamasensei** [jonataslaw/getx#2312] [Q&A]
Hi guys! I need help with auth middleware implementation using [`firebase_auth`](https://firebase.flutter.dev/docs/auth/overview/). I have tried so many attempts but none is working. The problem with my current implementation is that the user is redirecting to the login page everytime the app is reopened or the browser was refreshed.

My current middleware implementation looks like this, and what I know is the `currentUser` is always giving me a `null` result on page load.

`auth_middleware.dart`
```dart
class AuthMiddleware extends GetMiddleware {
  @override
  int? get priority => 0;

  @override
  RouteSettings? redirect(String? route) {
    return FirebaseAuth.instance.currentUser != null
        ? null
        : RouteSettings(name: Routes.AUTH);
  }
}
```

`app_pages.dart`
```dart
class AppPages {
  AppPages._();

  static const INITIAL = Routes.AUTH;

  static final routes = [
    GetPage(
      name: _Paths.HOME,
      page: () => HomeView(),
      binding: HomeBinding(),
      middlewares: [
        AuthMiddleware(),
      ],
    ),
    GetPage(
      name: _Paths.AUTH,
      page: () => AuthView(),
      binding: AuthBinding(),
      middlewares: [
        GuestMiddleware(),
      ],
    ),
  ];
}
```

I once tried to use `GetxService` but it sometimes throws a `navigation without context` error when I tried to run the app.
The `user` getter is also always giving me `null`. What I really get from this method is this happening because the `authStateChanges` stream is triggering the `ever` worker, eventhough the value is the same (`null`). I am wondering if I could do something to make the `ever` worker not triggering in at the very first time the stream is binded. Can we do that?

`auth_service.dart`
```dart
class AuthService extends GetxService {
  static AuthService get instance => Get.find();

  final _firebaseUser = Rxn<User>();

  model.User? get user {
    if (_firebaseUser() == null) return null;
    return model.User.fromFirebaseUser(_firebaseUser()!);
  }

  Future<AuthService> init() async {
    _firebaseUser.bindStream(FirebaseAuth.instance.authStateChanges());
    ever(_firebaseUser, _redirectIfUnauthenticated);
    return this;
  }

  void _redirectIfUnauthenticated(User? state) {
    if (state == null) Get.offAllNamed(Routes.AUTH);
  }
}
```

`auth_middleware.dart`
```dart
class AuthMiddleware extends GetMiddleware {
  @override
  int? get priority => 0;

  @override
  RouteSettings? redirect(String? route) {
    return AuthService.instance.user != null
        ? null
        : RouteSettings(name: Routes.AUTH);
  }
}
```

`main.dart`
```dart
void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await Get.putAsync(() => FirebaseService().init());
  await Get.putAsync(() => LoggerService().init());
  await Get.putAsync(() => AuthService().init());

  runApp(MyApp());
}
```

I really appreciate any kind of help from the community! ><

**Comment by elana-voss** [comment]
This works for with hot reload, as the user seems to be re-registered at some point and is not NULL.

```dart
sub = _firebaseAuth.authStateChanges().listen((User? user) {  
  if (user != null) {
    Get.offNamed( YOUR_ROUTE );
  }
});
```
